### ISSUE TITLE
Native prometheus metrics

### ISSUE THREAD

**Issue by jptosso** [corazawaf/coraza#1103]
Prometheus metrics are everywhere. We should support this as a first-class feature so our connectors can register and add WAF metrics to their applications.

This would benefit Cora Caddy and Wasm directly.

To do this, we could simply export a package that automatically registers the metrics or we could add helpers to add them at an application level.

**Comment by fzipi** [comment]
Related: https://github.com/corazawaf/coraza/issues/1008

**Comment by jcchavezs** [comment]
We first need to design the metrics to be tracked. That task would probably require people operating coraza e.g. @funkluk or @soujanyanmbri to propose interesting metrics and we all discuss them.

**Comment by jptosso** [comment]
It's hard to monitor all rule triggers. there are too many rules... what about having a filter to pick rules?
We also need to monitor disruptive actions
I found this reference https://github.com/XCiber/imperva-exporter/blob/d8c0c4af2652ce2f925e2f3ed0612d05346cf224/pkg/imperva/imperva.go

**Comment by jptosso** [comment]
Bringing this issue back to life, this would be a great addition for the experimental package so we can feed Caddy with metrics

**Comment by jycamier** [comment]
Currently, in my firsts tests with the Caddy pluging, I've created the following : 

| Metric | Type | Labels | Description |
|---|---|---|---|
| `coraza_waf_requests_total` | Counter | `method`, `host` | Total requests processed by the WAF |
| `coraza_waf_requests_blocked_total` | Counter | `method`, `action` | Requests blocked by the WAF |
| `coraza_waf_rules_triggered_total` | Counter | `severity` | Individual WAF rules triggered |

### Label values

- **`method`**: HTTP method (GET, POST, etc.)
- **`host`**: Server name from the Host header (port stripped)
- **`action`**: Interruption action (`deny`, `drop`, `redirect`)
- **`severity`**: Rule severity (`emergency`, `alert`, `critical`, `error`, `warning`, `notice`, `info`, `debug`)

### About rule ID

`rule_id` isn't in the labels due to possible hight cardinality issues.

**Comment by rikatz** [comment]
On CKO I have started to analyze what metrics for dataplane would look like: https://github.com/networking-incubator/coraza-kubernetes-operator/pull/399/changes#diff-39deb1d3d9a5e93cefe90f870546424ce539125cf8a69bce2db25b3e6e105cc8

I will bring the work to Coraza directly and spend some time this week analyzing how to register them, what is a good cardinality, if whatever claude proposed on this document makes sense.

**Comment by rikatz** [comment]
So here is something I was running on CKO, and that I think it could be a good bare minimum of metrics provided OOTB by Coraza (and eventually mutated by each engine as desired, enriching with rules, etc). Some of them are very close to what is proposed in https://github.com/corazawaf/coraza/issues/1103#issuecomment-4082250175

* `coraza_waf_requests_total` - Similar as proposed by @jycamier but adding the outcome as well - `pass, block, detect, redirect, error`
  *  I would probably remove the method and the host, as it doesn't matter as much as the outcome. host can grow cardinality a lot, method may not be useful IMHO and belongs to logs.
* `coraza_waf_blocked_requests_total` - No method, instead `category` and `severity`
* `coraza_waf_request_anomaly_score` - histogram - allow threshold tuning, false positive detection, etc (see https://github.com/networking-incubator/coraza-kubernetes-operator/blob/main/docs/driver-metrics-contract.md#coraza_waf_request_anomaly_score)
* Some extra helping metrics to debug plugin status, like "loaded rules" and "plugin loads"
~* `coraza_waf_rule_overrides` - Amount of rules that may be overridden. I don't have a strong opinion on this one yet, tho it would be nice to have some "alert" if users are overriding more rules than they should. This one is one where cardinality may be a problem~ - IMO this is a control plane and not a dataplane problem
* I am missing some histogram about processing time. At least a distribution of "amount of time spent on Coraza for the transaction processing" so one can follow if 'too many loaded rules means slower processing time'

What about ruleids and methods?

IMO those should belong to logs, and not to metrics. But we need a "good way" to make these logs visible and parseable (maybe json?) but I wont get into that for now, I think this is a nice to have followup.

**Comment by fzipi** [comment]
Late to this, but I went through @rikatz's [CKO contract](https://github.com/networking-incubator/coraza-kubernetes-operator/blob/main/docs/driver-metrics-contract.md) and @jycamier's set. Broad agreement, so I'll skip what's settled and raise one correction and one whole category I think is missing.

## The anomaly-score buckets don't fit CRS

This is the metric I care most about, and I think the proposed buckets make it unable to answer its main question.

From `crs-setup.conf.example`:

| Constant | Value |
|---|---|
| `tx.inbound_anomaly_score_threshold` | 5 |
| `tx.outbound_anomaly_score_threshold` | 4 |
| `critical / error / warning / notice` | 5 / 4 / 3 / 2 |

**One CRITICAL rule blocks.** The whole decision boundary sits at 5, and `error` is one point below it.

The CKO buckets are `0, 5, 10, 15, 20, 30, 40, 50, 75, 100`. Everything from 1 to 5 falls in a single bucket, so you cannot separate:

- score 0 — clean
- score 2 — one notice
- score 3 — one warning
- score 4 — one error, **one point from being blocked**
- score 5 — blocked

"How much legitimate traffic is one rule away from a block?" is the question you tune against, and those buckets can't express it. Once traffic reaches 20+ the exact value stops mattering — it's an attack either way.

I'd suggest `0, 1, 2, 3, 4, 5, 7, 10, 15, 20, 30, 50, +Inf`: integer-granular through the decision region, coarse above it. Same order of series count, much more signal. Inbound and outbound want separate histograms, since their thresholds differ (5 vs 4).

Also worth exporting `tx.inbound_anomaly_score_threshold` as a gauge, so dashboards compute the margin instead of hard-coding 5 and silently going wrong when an operator tunes it.

## Missing category: was the request fully inspected?

Every metric proposed so far measures **what the WAF decided**. None measures **whether the WAF saw the whole request** — and that is a different failure mode, invisible in every dashboard we've discussed.

Coraza drops input silently in several places today. From `internal/corazawaf/transaction.go` on `main`:

```
:773  tx.debugLogger.Warn().Msg("skipping get request argument, over limit")
:782  tx.debugLogger.Warn().Msg("skipping post request argument, over limit")
:791  tx.debugLogger.Warn().Msg("skipping path request argument, over limit")
:804  tx.debugLogger.Warn().Msg("skipping response argument, over limit")
```

Debug log only. No variable, no counter. If `SecArgumentsLimit` is reached, arguments are discarded and the rules never see them — and nobody watching a dashboard can tell. The same shape applies to body truncation at `SecRequestBodyLimit` with `ProcessPartial`, and to a body processor that fails to parse: rules then run against a value that isn't what the origin will parse.

Proposed:

| Metric | Type | Labels |
|---|---|---|
| `coraza_waf_input_truncated_total` | counter | `stage` = `args_get`, `args_post`, `args_path`, `args_response`, `request_body`, `response_body` |
| `coraza_waf_body_processor_total` | counter | `processor` = `json`,`xml`,`multipart`,`urlencoded`; `result` = `ok`,`error` |
| `coraza_waf_rule_errors_total` | counter | none |

Cheap to add — the checks already exist, and `REQBODY_ERROR`, `REQBODY_PROCESSOR_ERROR`, `MULTIPART_STRICT_ERROR` and `URLENCODED_ERROR` are already computed. These are correctness metrics, not performance ones: a rise means requests are being evaluated on partial input, which is the interesting thing to alert on and something an attacker can deliberately induce.

## Two smaller suggestions

**Use the rule ID range as the attack category.** In CRS the range *is* the taxonomy — 913 scanner, 920/921 protocol, 930 LFI, 931 RFI, 932 RCE, 933 PHP, 934 generic, 941 XSS, 942 SQLi, 943 fixation, 944 Java. That's ~15 fixed values: category attribution with no tag parsing, no top-N slots and no `other` overflow bucket. It also separates scanning (913-heavy) from exploitation, which you want to alert on differently. I think this beats a 200-slot `rule_id` label, and it agrees with the thread's instinct that rule IDs belong in logs.

**Label the latency histogram and interruptions by phase.** `request_headers` / `request_body` / `response_headers` / `response_body`. Response-body inspection is usually the expensive phase and operators rarely know it; and an interruption in phase 4 means the response was already generated, which is worth seeing separately.

## The design question underneath all this

Coraza core has no knowledge of CRS. Anomaly score, paranoia level and attack category are rule-set concepts living in `TX` variables, so either Coraza ships a generic mechanism — "export these `TX` variables as metrics, with these buckets" — or every connector reimplements the CRS-aware half. I lean toward a generic hook plus a small CRS-flavoured preset, but that seems worth settling before anyone writes the collector.

Happy to prototype the engine-level metrics (the truncation/parser set and phase timings) if there's agreement on shape — those are the ones that need no rule-set knowledge at all.

**Comment by funkluk** [comment]
> ...
> ## The design question underneath all this
> Coraza core has no knowledge of CRS. Anomaly score, paranoia level and attack category are rule-set concepts living in TX variables, so either Coraza ships a generic mechanism — "export these TX variables as metrics, with these buckets" — or every connector reimplements the CRS-aware half. I lean toward a generic hook plus a small CRS-flavoured preset, but that seems worth settling before anyone writes the collector.
> ...

I think this is what would be very useful. This would allow any Coraza-/CRS-plugins, in addition to the CRS itself, to generate metrics as required.
